### DISCUSSION TITLE
MetaAgents and agent activation based on networks

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#2539] [Ideas]
With AgentSet being firmly embedded and excellent features like `groupby` I think we can leverage these to allow the dynamic creation of meta agents fairly easily.  I already started some testing and have a path forward but wanted to get thoughts to see if it is valid/ worth the effort. 

**Conceptual Foundation:** Complex systems are adaptive networks. Using the brain as a metaphor, when we receive stimulus this activates groups of neurons to produce thoughts and behaviors. At any given time we have active and dormant neural networks. 

**Question:** How can we provide features that effectively keep sets of linked agents, which given stimulus in the model (ideally endogenously) activates these networks?

**Proposed Approach:** Add a function to AgentSet that dynamically creates new agent types that are classes of these linked agents.

**Slightly More than Pseudo Code:**

```
def create_metaagent(self, new_agent_class_name, list_of_agents, **unique_attributes_behaviors):
      #Get agent types
      agent_types =(agent.__class__ for agent in list_of_agents)      
      #Create new class
      meta_agent_class = type(new_agent_class_name, 
                                      (agent_types,), 
                                       **list_of_belonging_agents_as_an_attribute and unique_attributes_behaviors)
      #Add to globals
      globals()[new_agent_class_name]   
```

Some optional features may be a `kwarg` ensuring levels named as `ints` but could represent individuals, families, neighborhoods or team, league etc. A `kwarg` on whether on not any individual agent could be in multiple groups at the same level. Potentially, if not agents are restricted to one meta agent at any level this would facilitate for multithreading in Python 3.13. 

Curious on thoughts?

**Comment by EwoutH** [comment]
Very interesting topic! I once did a project in which households were very important, Households made car purchase and sharing decisions together. This concept seems to help/enable that.

One initial question: From my understanding, emergence in complex systems is fundamentally about patterns and behaviors arising naturally from the interactions of lower-level components, without being explicitly programmed or structured from above. This proposal seems to suggest that it can be build explicitly. Could you expand a little on what your vision is in this regard?

**Reply by tpike3** [reply]
@EwoutH  Sorry for the delay, thanks for the question. I would say this allows for bottom up emergence and policy at multiple levels. 

Generally, I would say complex systems often focus on bottom up emergence for a wide variety of scientific and historical reasons. However, to quote David Krakauer from SFI -- Complex systems are adaptive networks. So, societies emerge from individuals  but then those governing organizations impose laws that shape individuals behavior. Similarly, entities within an ecosystem create an emergent ecosystem, but that ecosystem then shapes the fitness landscape on which entities within that ecosystem adapt. For me complex systems is about the flow of dynamics across networks and associated adaptions and subsequent flows. 

The intent of this is to provide more options for both bottom up emergence, but also for the emergence of meta agents that can shape policies.

This paper is pretty rough as I did it several years ago, but it was my first foray into this topic. 

https://arxiv.org/pdf/1904.08315

You can also see I didn't quite get sugarscape with traders right, but later for the Complexity Explorer tutorial @rht and I found the issues.

**Reply by EwoutH** [reply]
Interesting, thanks! Will dive into it.

**Comment by tpike3** [comment]
See #2561 for initial implementation -- to start this currently does not allow for agents to belong multiple meta-agents due to the exponential growth of potential combinations, however future iterations can involve such options

**Comment by EwoutH** [comment]
## MetaAgent as Agent + AgentSet?
Looking at the current `MetaAgent` implementation alongside `AgentSet`, there's significant overlap. Both are fundamentally "a group of agents" with iteration, containment, and operations over members.

The key distinction is that a `MetaAgent` is a group that (also) is itself an agent: it has a `unique_id`, participates in the model, and can be scheduled/activated.

So would it make sense to make `MetaAgent` a subclass of both `Agent` and `AgentSet`? The core would be quite simple:

```python
class MetaAgent(Agent, AgentSet):
    """An AgentSet that is also an Agent."""
    
    membership_name: str = "meta_agent"
    
    def __init__(self, model, agents=None, name=None):
        Agent.__init__(self, model)
        AgentSet.__init__(self, agents or [], random=model.random)
        self.name = name or self.__class__.__name__
        for agent in self:
            setattr(agent, self.membership_name, self)
    
    def add(self, agent):
        super().add(agent)
        setattr(agent, self.membership_name, self)
    
    def discard(self, agent):
        super().discard(agent)
        if getattr(agent, self.membership_name, None) is self:
            setattr(agent, self.membership_name, None)
```

This makes all `AgentSet` methods work directly on a `MetaAgent`:

```python
class Household(MetaAgent):
    membership_name = "household"
    
    @property
    def budget(self):
        return self.agg("wealth", sum)

household = Household(model, {alice, bob})

# AgentSet methods
household.agg("age", [min, max])
household.select(filter_func=lambda a: a.employed)
household.groupby("role").count()
household.do("step")

# Agent methods
household.unique_id
household.step()

# Back-reference
alice.household.budget
```

### Questions for discussion
1. Any concerns with multiple inheritance from `Agent` and `AgentSet`?
2. Should `membership_name` back-references be opt-in rather than default?
3. Do we want multi-membership support in the base class or as an optional mixin?

Curious what you think!

**Reply by edgeofchaos42** [reply]
I agree I think sets is the way to go. 

A colleague (@DemetersSon83) and I used to discuss ideas that if complex systems are adaptive networks and deep learning is  structured adaptive networks, that at some point the two will converge. I think hypernetworks (based on set theory instead of graph theory) and topology will provide the math.

In short, I think the meta-agent challenge is figuring out the set management of a multi-level adaptive network of activating and deactivating dynamically changing sets of agents. That is the module we need to build. I think meta-agent is a rough hewn start.

**Reply by DemetersSon83** [reply]
@edgeofchaos42 It would be interesting to build a simple contagion toy model using the MetaAgent framework described here on a hypergraph topology as a proof of concept.

**Reply by tpike3** [reply]
That would be... I'll ping you to discuss

**Comment by EwoutH** [comment]
For the examples, maybe we can update some:
- In Boid flockers, a flock could be a meta agent
- In conways game of life, a group of cells (like a walker) could be a meta agent

Households are generally also very strong candidates for meta agents. They could share budgets, beliefs, transportation means, or could make joint decisions on holiday destinations, groceries, etc.

**Reply by quaquel** [reply]
I would not touch those examples. Keep them simple and classic. Also, in both cases, the meta agent has no additional behavior. Households or coalitions are better examples. For now, however, I believe the next steps are to improve the code and the alliance-formation example.

**Reply by edgeofchaos42** [reply]
I concur, we should keep the classics, classic. However, maybe we will be able to grow a whole new set of meta-agent seminal models 😀

**Reply by EwoutH** [reply]
Maybe not in the main repo, but let’s go wild on other places.

What if a flock has a shared direction property, on which agents base their direction, instead of (only) on the bird direct in front of them?

**Reply by tpike3** [reply]
I would be curious of the impacts of that... if I am understanding correctly the classis boids model is the idea complex emergent behavior from simple rules and now their would be a collective consciousness if you will.. Would it have an impact, how would each behave under different scenarios. That could be pretty interesting.

**Reply by DemetersSon83** [reply]
The Boids model is an "upward causation" model where the flock behavior property is emergent. Perhaps I'm misunderstanding, but if you (@EwoutH) are suggesting a shared direction property that is predefined, that would be a "downward causation" model and so would not really be a model of emergence. I think the standard Boids model would be fairly straightforward to model and would be interesting for sure.

**Reply by EwoutH** [reply]
Interesting point. But as long as every boid still determines their own direction, it's still emergence, right?

Having a model that can do:
- Following the closest (current behavior)
- Following the group direction (meta-agent)
- Following weighted (distance based, weighted/decay up to max)

and be able to compare that, could be interesting. ABM is about interactions, and emergence is just a specific case of that (while an interesting one).
